prometric/src/summary/generic.rs

```diff
@@ -0,0 +1,317 @@
+        Self {
+            common_opts: Opts::new(name, help),
+            summary_opts: summary,
+            quantiles: Vec::from(DEFAULT_QUANTILES),
+        }
+    }
+
+    /// See [`Opts::const_labels`]
+    pub fn const_labels(mut self, const_labels: HashMap<String, String>) -> Self {
+        self.common_opts = self.common_opts.const_labels(const_labels);
+        self
+    }
+
+    /// See [`Opts::variable_labels`]
+    pub fn variable_labels(mut self, variable_labels: Vec<String>) -> Self {
+        self.common_opts = self.common_opts.variable_labels(variable_labels);
+        self
+    }
+
+    /// Configure the quantiles to use when creating a prometheus protobuf summary
+    pub fn quantiles<B: Into<Vec<f64>>>(self, quantiles: B) -> Self {
+        Self { quantiles: quantiles.into(), ..self }
+    }
+}
+
+/// Uses the configured [`SummaryProvider`] `P` to collect observations and compute quantiles
+///
+/// Main purpose is to wrap over the summary to convert it into a [`prometheus::proto::Summary`]
+#[derive(Debug)]
+pub struct GenericSummary<P> {
+    label_pairs: Vec<pp::LabelPair>,
+
+    provider: P,
+
+    quantiles: Vec<f64>,
+}
+
+impl<P: NonConcurrentSummaryProvider> GenericSummary<P> {
+    pub fn new<V: AsRef<str>>(
+        opts: &SummaryOpts<P::Opts>,
+        label_values: &[V],
+    ) -> prometheus::Result<Self> {
+        let desc = opts.common_opts.describe()?;
+        let label_pairs = make_label_pairs(&desc, label_values)?;
+
+        Ok(Self {
+            label_pairs,
+            provider: P::new_provider(&opts.summary_opts),
+            quantiles: opts.quantiles.clone(),
+        })
+    }
+
+    /// Make a snapshot of the current summary state exposed as a Protobuf struct
+    pub fn proto(&self) -> pp::Summary {
+        let snapshot = self.provider.snapshot();
+        let mut summary = pp::Summary::default();
+
+        summary.set_sample_sum(snapshot.sample_sum());
+        summary.set_sample_count(snapshot.sample_count());
+
+        let mut quantiles = Vec::with_capacity(self.quantiles.len());
+        for quantile in self.quantiles.iter().cloned() {
+            let mut q = pp::Quantile::default();
+            q.set_quantile(quantile);
+
+            // TODO: signal that this value was not computable if == None
+            let Some(val) = snapshot.quantile(quantile) else { continue };
+            q.set_value(val);
+
+            quantiles.push(q);
+        }
+
+        summary.set_quantile(quantiles);
+
+        summary
+    }
+}
+
+impl<P> Deref for GenericSummary<P> {
+    type Target = P;
+
+    fn deref(&self) -> &Self::Target {
+        &self.provider
+    }
+}
+
+#[derive(Clone)]
+/// NewType over [`GenericSummaryImpl`]
+///
+/// Uses `Arc` to ensure clones refer to the same data.
+/// This is becuase [ `prometheus::core::MetricVec` ] will clone the metric each time it is to be
+/// accessed, even when inserting new data
```

Comment:
```suggestion
/// NewType over [`GenericSummaryImpl`]
///
/// Uses `Arc` to ensure clones refer to the same data.
/// This is because [ `prometheus::core::MetricVec` ] will clone the metric each time it is to be
/// accessed, even when inserting new data.
#[derive(Clone)]
```